login-utils/vipw.c

```diff
@@ -94,7 +94,7 @@ static void copyfile(int from, int to)
 	char buf[8 * 1024];
 
 	while ((nr = read(from, buf, sizeof(buf))) > 0)
-		for (off = 0; off < nr; nr -= nw, off += nw)
+		for (off = 0; nr > 0; nr -= nw, off += nw)
```

Comment:
Maybe this function could be simplified, something like:

    static void copyfile(int from, int to)
    {
            char buf[8 * 1024];
            ssize_t sz;
    
            while (0 < (sz = read(from, buf, sizeof(buf))))
                    if (write(to, buf, sz) != sz)
                            pw_error(tmp_file, 1, 1);
    
    #ifdef HAVE_EXPLICIT_BZERO
            explicit_bzero(buf, sizeof(buf));
    #endif
    }

Reply:
Yeah, this is how I'd typically do it. Retrying partial writes sounds more robust, but I'm not sure if it actually is.

Reply:
It turns out this approach makes the correct error handling a bit more complicated. We also don't get errno, so we must output a generic "write error" message.
```c
static void copyfile(int from, int to)
{
	ssize_t nr, nw;
	char buf[8 * 1024];

	while ((nr = read(from, buf, sizeof(buf))) > 0)
		if ((nw = write(to, buf, nr)) != nr) {
			if (nw < 0)
				pw_error(tmp_file, 1, 1);
			warnx(_("%s: write error"), tmp_file);
			pw_error(NULL, 0, 1);
		}

	if (nr < 0)
		pw_error(orig_file, 1, 1);
#ifdef HAVE_EXPLICIT_BZERO
	explicit_bzero(buf, sizeof(buf));
#endif
}
```
(commit https://github.com/DankRank/util-linux/commit/5495d873d02d788516bcfa7e33771641d972eb54)

Either version is fine with me.